core/src/main/java/hudson/PluginManager.java

```diff
@@ -1337,6 +1337,7 @@ public HttpResponse doPlugins() {
                 pluginInfo.put("dependencies", Collections.emptyMap());
             }
             response.add(pluginInfo);
+            allPlugins.put(plugin.getShortName(), pluginInfo);
         }
         for (UpdateSite site : Jenkins.getActiveInstance().getUpdateCenter().getSiteList()) {
             for (UpdateSite.Plugin plugin: site.getAvailables()) {
```

Comment:
Without the patch a new JSONObject is created for every entry below, even if the plugin has been added to the response before

Reply:
Ah yes, sorry.

Maybe the correct behaviour is to _not_ add the `pluginInfo` to the response in the second loop if it has already been added in the first loop?

I note that the `"excerpt"` and `"site"` keys are only populated in `pluginInfo` in the second loop. I guess these are only relevant for plugins that aren't installed.

Reply:
Sorry, missed the comment

> Maybe the correct behaviour is to _not_ add the `pluginInfo` to the response in the second loop if it has already been added in the first loop?

Looks like a plan

> I note that the `"excerpt"` and `"site"` keys are only populated in `pluginInfo` in the second loop. I guess these are only relevant for plugins that aren't installed

Hard to say. I would append the data to the response if it does not require much time